File: apps/web/src/utils/occurrence-generation.ts
import type { OccurrenceStatus } from "@/types/occurrence";
import type { TransactionKind, TransactionRecurrenceMode } from "@/types/transaction";
import {
  createMonthlyOccurrence,
  formatDateValue,
  getTodayDateValue,
  parseDateValue,
} from "@/utils/recurring-transactions";

export type GeneratedOccurrenceExtension = GeneratedOccurrence;

export type GenerateOccurrenceExtensionInput = {
  recurrenceDay: number;
  amount: number;
  nextInstallmentIndex: number;
  /** Primeira DueDate a gerar (mês seguinte ao último existente) como "YYYY-MM-DD" */
  fromDate: string;
  /** Data limite — até quando gerar (hoje + 12 meses) como "YYYY-MM-DD" */
  horizonEnd: string;
};

export type GeneratedOccurrence = {
  dueDate: string;
  amount: number;
  installmentIndex: number | null;
  status: OccurrenceStatus;
};

export type GenerateOccurrencesInput = {
  transactionKind: TransactionKind;
  amount: number;
  transactionDate: string | null;
  installmentCount: number | null;
  installmentStartDate: string | null;
  recurrenceDay: number | null;
  recurrenceStartDate: string | null;
  recurrenceMode: TransactionRecurrenceMode | null;
  recurrenceEndDate: string | null;
  recurrenceMonths: number | null;
};

function addMonthsToDate(date: Date, monthOffset: number) {
  const totalMonths = date.getMonth() + monthOffset;
  const year = date.getFullYear() + Math.floor(totalMonths / 12);
  const monthIndex = ((totalMonths % 12) + 12) % 12;

  return { year, monthIndex };
}

function buildOccurrenceDate(anchorDate: Date, dayOfMonth: number, monthOffset: number) {
  const { year, monthIndex } = addMonthsToDate(anchorDate, monthOffset);

  return createMonthlyOccurrence(year, monthIndex, dayOfMonth);
}

/**
 * Réplica em TypeScript das regras exatas de
 * Finly.Application/Services/OccurrenceGenerationService.cs — Single gera 1 occurrence,
 * Installment gera N mensais a partir da data informada, Recurring gera até a condição de
 * parada (until-date/for-months/12 meses se indefinido). Status Paid/Pending é decidido do
 * mesmo jeito: DueDate <= hoje. É a fonte ÚNICA de geração de occurrences no modo local
 * (use-local-finance.ts persiste o resultado de verdade) e também alimenta o preview de
 * simulação (use-impact-simulation.ts, para os dois modos) — não duplicar esta lógica.
 */
export function generateOccurrences(input: GenerateOccurrencesInput): GeneratedOccurrence[] {
  const today = parseDateValue(getTodayDateValue())!;

  function buildOccurrence(dueDate: Date, installmentIndex: number | null): GeneratedOccurrence {
    return {
      dueDate: formatDateValue(dueDate),
      amount: input.amount,
      installmentIndex,
      status: dueDate <= today ? "paid" : "pending",
    };
  }

  if (input.transactionKind === "installment-template") {
    const startDate = parseDateValue(input.installmentStartDate);

    if (!startDate || !input.installmentCount) {
      return [];
    }

    const dayOfMonth = startDate.getDate();
    const occurrences: GeneratedOccurrence[] = [];

    for (let index = 0; index < input.installmentCount; index += 1) {
      const dueDate = buildOccurrenceDate(startDate, dayOfMonth, index);
      occurrences.push(buildOccurrence(dueDate, index + 1));
    }

    return occurrences;
  }

  if (input.transactionKind === "recurring-template") {
    const startDate = parseDateValue(input.recurrenceStartDate);

    if (!startDate) {
      return [];
    }

    const dayOfMonth = input.recurrenceDay ?? startDate.getDate();
    const mode = input.recurrenceMode ?? "indefinite";
    const endDate = parseDateValue(input.recurrenceEndDate);
    const occurrences: GeneratedOccurrence[] = [];
    let monthOffset = 0;

    while (true) {
      const dueDate = buildOccurrenceDate(startDate, dayOfMonth, monthOffset);

      if (mode === "until-date") {
        if (endDate && dueDate > endDate) {
          break;
        }
      } else if (mode === "for-months") {
        if (monthOffset >= (input.recurrenceMonths ?? 0)) {
          break;
        }
      } else if (monthOffset >= 12) {
        break;
      }

      occurrences.push(buildOccurrence(dueDate, monthOffset + 1));
      monthOffset += 1;
    }

    return occurrences;
  }

  const transactionDate = parseDateValue(input.transactionDate) ?? today;

  return [buildOccurrence(transactionDate, null)];
}

/**
 * Gera occurrences adicionais para uma recorrência indefinida cujo horizonte está se esgotando.
 * Espelha OccurrenceGenerationService.GenerateExtension do backend.
 * fromDate é o primeiro mês a gerar (mês seguinte ao último existente).
 * InstallmentIndex continua sequencialmente a partir de nextInstallmentIndex.
 */
export function generateOccurrenceExtension(
  input: GenerateOccurrenceExtensionInput,
): GeneratedOccurrenceExtension[] {
  const today = parseDateValue(getTodayDateValue())!;
  const fromDate = parseDateValue(input.fromDate);
  const horizonEnd = parseDateValue(input.horizonEnd);

  if (!fromDate || !horizonEnd) {
    return [];
  }

  const occurrences: GeneratedOccurrenceExtension[] = [];
  let installmentIndex = input.nextInstallmentIndex;
  let monthOffset = 0;

  while (true) {
    const dueDate = buildOccurrenceDate(fromDate, input.recurrenceDay, monthOffset);

    if (dueDate > horizonEnd) {
      break;
    }

    occurrences.push({
      dueDate: formatDateValue(dueDate),
      amount: input.amount,
      installmentIndex,
      status: dueDate <= today ? "paid" : "pending",
    });

    installmentIndex += 1;
    monthOffset += 1;
  }

  return occurrences;
}

File: apps/mobile/lib/features/transactions/data/occurrence_generation.dart
import '../../../shared/models/enums.dart';
import 'transaction_form_input.dart';

/// Réplica em Dart de apps/web/src/utils/occurrence-generation.ts
/// (`generateOccurrences`), que por sua vez replica
/// Finly.Application/Services/OccurrenceGenerationService.cs.
///
/// Usada pelo preview de "Simular impacto" (não persiste) **e** pelo modo sem
/// conta (LocalFinanceStore persiste o resultado de verdade). Single = 1
/// ocorrência; Installment = N mensais; Recurring = até a condição de parada
/// (até-data / por-meses / 12 meses se indefinido). Status Paid/Pending decidido
/// por `dueDate <= hoje`, igual ao backend.
class GeneratedOccurrence {
  GeneratedOccurrence({
    required this.dueDate,
    required this.amount,
    required this.installmentIndex,
    required this.paid,
  });

  final DateTime dueDate;
  final double amount;
  final int? installmentIndex;
  final bool paid;
}

DateTime _monthlyOccurrence(DateTime anchor, int dayOfMonth, int monthOffset) {
  final totalMonths = anchor.month - 1 + monthOffset;
  final year = anchor.year + (totalMonths / 12).floor();
  final monthIndex = totalMonths % 12;
  final lastDay = DateTime(year, monthIndex + 2, 0).day;
  final day = dayOfMonth < lastDay ? dayOfMonth : lastDay;
  return DateTime(year, monthIndex + 1, day, 12);
}

List<GeneratedOccurrence> generateOccurrences(TransactionFormInput input) {
  final now = DateTime.now();
  final today = DateTime(now.year, now.month, now.day, 12);

  GeneratedOccurrence build(DateTime dueDate, int? index) => GeneratedOccurrence(
        dueDate: dueDate,
        amount: input.amount,
        installmentIndex: index,
        paid: !dueDate.isAfter(today),
      );

  switch (input.kind) {
    case ContractKind.installment:
      final start = DateTime.tryParse(input.installmentStartDate);
      final count = input.installmentCount ?? 0;
      if (start == null || count < 1) return const [];
      return [
        for (var i = 0; i < count; i++)
          build(_monthlyOccurrence(start, start.day, i), i + 1),
      ];

    case ContractKind.recurring:
      final start = DateTime.tryParse(input.recurrenceStartDate);
      if (start == null) return const [];
      final day = input.recurrenceDay ?? start.day;
      final end = input.recurrenceEndDate == null
          ? null
          : DateTime.tryParse(input.recurrenceEndDate!);
      final result = <GeneratedOccurrence>[];
      var offset = 0;
      while (true) {
        final due = _monthlyOccurrence(start, day, offset);
        if (input.recurrenceMode == RecurrenceMode.untilDate) {
          if (end != null && due.isAfter(end)) break;
        } else if (input.recurrenceMode == RecurrenceMode.forMonths) {
          if (offset >= (input.recurrenceMonths ?? 0)) break;
        } else if (offset >= 12) {
          break;
        }
        result.add(build(due, offset + 1));
        offset += 1;
      }
      return result;

    case ContractKind.single:
      final date = DateTime.tryParse(input.transactionDate) ?? today;
      return [build(date, null)];
  }
}

/// Réplica de `generateOccurrenceExtension` (occurrence-generation.ts) — gera
/// ocorrências adicionais para uma recorrência indefinida cujo horizonte está
/// se esgotando. Espelha `TransactionService.ExtendIndefiniteRecurrencesIfNeeded`
/// do backend. Usada só no modo sem conta ao carregar (ver LocalFinanceStore).
List<GeneratedOccurrence> generateOccurrenceExtension({
  required int recurrenceDay,
  required double amount,
  required int nextInstallmentIndex,

  /// Primeira DueDate a gerar (mês seguinte ao último existente), "YYYY-MM-DD".
  required DateTime fromDate,

  /// Até quando gerar (hoje + 12 meses).
  required DateTime horizonEnd,
}) {
  final now = DateTime.now();
  final today = DateTime(now.year, now.month, now.day, 12);

  final result = <GeneratedOccurrence>[];
  var index = nextInstallmentIndex;
  var offset = 0;
  while (true) {
    final due = _monthlyOccurrence(fromDate, recurrenceDay, offset);
    if (due.isAfter(horizonEnd)) break;
    result.add(GeneratedOccurrence(
      dueDate: due,
      amount: amount,
      installmentIndex: index,
      paid: !due.isAfter(today),
    ));
    index += 1;
    offset += 1;
  }
  return result;
}

/// Impacto no saldo atual (apenas as ocorrências que nasceriam **pagas**) —
/// o mesmo que o web mostra no card de simulação.
double simulationPaidDelta(TransactionFormInput input) {
  final signed = input.type == TransactionType.income ? 1 : -1;
  return generateOccurrences(input)
      .where((o) => o.paid)
      .fold<double>(0, (sum, o) => sum + signed * o.amount);
}
